src/Microsoft.Data/ArrowStringColumn.cs

```diff
@@ -0,0 +1,348 @@
+        }
+
+        private int GetNullCount(long startIndex, int numberOfRows)
+        {
+            int nullCount = 0;
+            for (long i = startIndex; i < numberOfRows; i++)
+            {
+                if (!IsValid(i))
+                    nullCount++;
+            }
+            return nullCount;
+        }
+
+
+        public override Apache.Arrow.Array AsArrowArray(long startIndex, int numberOfRows)
+        {
+            int offsetsBufferIndex = GetBufferIndexContainingRowIndex(ref startIndex);
+            ArrowBuffer dataBuffer = numberOfRows == 0 ? ArrowBuffer.Empty : new ArrowBuffer(_dataBuffers[offsetsBufferIndex].ReadOnlyMemory);
+            ArrowBuffer offsetsBuffer = numberOfRows == 0 ? ArrowBuffer.Empty : new ArrowBuffer(_offsetsBuffers[offsetsBufferIndex].ReadOnlyMemory);
+            ArrowBuffer nullBuffer = numberOfRows == 0 ? ArrowBuffer.Empty : new ArrowBuffer(_nullBitMapBuffers[offsetsBufferIndex].ReadOnlyMemory);
+            int nullCount = GetNullCount(startIndex, numberOfRows);
+            return new StringArray(numberOfRows, offsetsBuffer, dataBuffer, nullBuffer, nullCount, (int)startIndex);
+        }
+
+        public override BaseColumn Sort(bool ascending = true) => throw new NotSupportedException();
+
+        public override BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false, long numberOfNullsToAppend = 0)
+        {
+            ArrowStringColumn clone;
+            if (!(mapIndices is null))
+            {
+                if (mapIndices.DataType != typeof(long))
+                    throw new ArgumentException(Strings.MismatchedValueType + " PrimitiveColumn<long>", nameof(mapIndices));
+                if (mapIndices.Length > Length)
+                    throw new ArgumentException(Strings.MapIndicesExceedsColumnLenth, nameof(mapIndices));
+                clone = Clone(mapIndices as PrimitiveColumn<long>, invertMapIndices);
+            }
+            else
+            {
+                clone = Clone();
+            }
+            for (long i = 0; i < numberOfNullsToAppend; i++)
+            {
+                clone.Append(null);
+            }
+            return clone;
+        }
+
+        private ArrowStringColumn Clone(PrimitiveColumn<long> mapIndices = null, bool invertMapIndex = false)
+        {
+            ArrowStringColumn ret = new ArrowStringColumn(Name);
+            if (mapIndices is null)
+            {
+                for (long i = 0; i < Length; i++)
+                {
+                    ret.Append(this[i]);
+                }
+            }
+            else
+            {
+                if (mapIndices.Length > Length)
+                    throw new ArgumentException(Strings.MapIndicesExceedsColumnLenth, nameof(mapIndices));
+                if (invertMapIndex == false)
+                {
+                    for (long i = 0; i < mapIndices.Length; i++)
+                    {
+                        ret.Append(this[(long)mapIndices[i]]);
```

Comment:
Why is the cast to `long` here necessary?

Reply:
Fixed. My bad.